service/dap/server.go

```diff
@@ -1196,94 +1200,128 @@ func (s *Session) onSetBreakpointsRequest(request *dap.SetBreakpointsRequest) {
 	clientPath := request.Arguments.Source.Path
 	serverPath := s.toServerPath(clientPath)
 
-	// According to the spec we should "set multiple breakpoints for a single source
-	// and clear all previous breakpoints in that source." The simplest way is
-	// to clear all and then set all. To maintain state (for hit count conditions)
-	// we want to amend existing breakpoints.
-	//
+	// Get all existing breakpoints that match for this source.
+	sourceRequestPrefix := fmt.Sprintf("sourceBp Path=%q ", request.Arguments.Source.Path)
+
+	breakpoints := s.setBreakpoints(sourceRequestPrefix, len(request.Arguments.Breakpoints), func(i int) *bpMetadata {
+		want := request.Arguments.Breakpoints[i]
+		return &bpMetadata{
+			name:         fmt.Sprintf("%s Line=%d Column=%d", sourceRequestPrefix, want.Line, want.Column),
+			condition:    want.Condition,
+			hitCondition: want.HitCondition,
+			logMessage:   want.LogMessage,
+		}
+	}, func(i int) (*bpLocation, error) {
+		want := request.Arguments.Breakpoints[i]
+		return &bpLocation{
+			File: serverPath,
+			Line: want.Line,
+		}, nil
+	})
+
+	response := &dap.SetBreakpointsResponse{Response: *newResponse(request.Request)}
+	response.Body.Breakpoints = breakpoints
+
+	s.send(response)
+}
+
+type bpMetadata struct {
+	name         string
+	condition    string
+	hitCondition string
+	logMessage   string
+}
+
+type bpLocation struct {
+	File  string
+	Line  int
+	Addr  uint64
+	Addrs []uint64
+}
+
+// setBreakpoints is a helper function for setting source, function and instruction
+// breakpoints. It takes the prefix of the name for all breakpoints that should be
+// included, the total number of breakpoints, and functions for computing the metadata
+// and the location. The location is computed separately because this may be more
+// expensive to compute and may not always be necessary.
+func (s *Session) setBreakpoints(prefix string, totalBps int, metadataFunc func(i int) *bpMetadata, locFunc func(i int) (*bpLocation, error)) []dap.Breakpoint {
 	// See https://github.com/golang/vscode-go/issues/163 for details.
 	// If a breakpoint:
 	// -- exists and not in request => ClearBreakpoint
 	// -- exists and in request => AmendBreakpoint
 	// -- doesn't exist and in request => SetBreakpoint
 
-	// Get all existing breakpoints that match for this source.
-	sourceRequestPrefix := fmt.Sprintf("sourceBp Path=%q ", request.Arguments.Source.Path)
-	existingBps := s.getMatchingBreakpoints(sourceRequestPrefix)
+	// Get all existing breakpoints matching the prefix.
+	existingBps := s.getMatchingBreakpoints(prefix)
 	bpAdded := make(map[string]struct{}, len(existingBps))
```

Comment:
What are bpAdded? The name is confusing.

Reply:
Added a comment to explain.